### PR TITLE
Unparsing: fix joins unparsing when table scan contains filters

### PR DESCRIPTION
**Pull request by sgrebnov**
## Which issue does this PR close?

Currently all `TableScan` filters and joins w/o dedicated projection incorrectly apply after join statement which leads to incorrect result. 

Original query (TPC-H 4)
```sql
select
            c_custkey,
            count(o_orderkey)
        from
            customer left join orders on
                        c_custkey = o_custkey
                    and o_comment not like '%special%requests%'
        group by
            c_custkey
```

Unparsed. You can see that filter (`where`) is incorrectly applied after join (not before via subquery or during join). The difference is that original query contains customer records where there is no matching order and the second version does not.

```sql
select
    "customer"."c_custkey",
    count("orders"."o_orderkey")
  from
    "customer"
  left join "orders" on
    ("customer"."c_custkey" = "orders"."o_custkey")
  where
    "orders"."o_comment" not like '%special%requests%'
  group by
    "customer"."c_custkey"
  order by "customer"."c_custkey" asc
```

This PR improves this by determining if join is simple table scan (not a subquery that will be translated to full subquery `(SELECT .. from ..)`) so filters can must be applied to join directly. In case of full subquery filters will be applied in subquery and this works correct. This fixes TPC-H 12 (filters on left join were missing/not applied at all) and TPC-H 13 (filter applied after join instead of during/before as showed above)

```console
SELECT "customer"."c_custkey", count("orders"."o_orderkey") FROM "customer" LEFT JOIN "orders" ON (("customer"."c_custkey" = "orders"."o_custkey") AND "orders"."o_comment" NOT LIKE '%special%requests%') GROUP BY "customer"."c_custkey" 


sql> explain select
            c_custkey,
            count(o_orderkey)
        from
            customer left join orders on
                        c_custkey = o_custkey
                    and o_comment not like '%special%requests%'
        group by
            c_custkey;
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | BytesProcessedNode                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
|               |   Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|               |  Projection: customer.c_custkey, count(orders.o_orderkey)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |   Aggregate: groupBy=[[customer.c_custkey]], aggr=[[count(orders.o_orderkey)]]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
|               |     Left Join:  Filter: customer.c_custkey = orders.o_custkey                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
|               |       TableScan: customer                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|               |       TableScan: orders, full_filters=[orders.o_comment NOT LIKE Utf8("%special%requests%")]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| physical_plan | BytesProcessedExec                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
|               |   SchemaCastScanExec                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
|               |     VirtualExecutionPlan name=postgres compute_context=host=Tcp("localhost"),port=6432,db=tpch_sf1,user=postgres, sql=SELECT customer.c_custkey, count(orders.o_orderkey) FROM customer LEFT JOIN orders ON ((customer.c_custkey = orders.o_custkey) AND orders.o_comment NOT LIKE '%special%requests%') GROUP BY customer.c_custkey rewritten_sql=SELECT "customer"."c_custkey", count("orders"."o_orderkey") FROM "customer" LEFT JOIN "orders" ON (("customer"."c_custkey" = "orders"."o_custkey") AND "orders"."o_comment" NOT LIKE '%special%requests%') GROUP BY "customer"."c_custkey" |
|               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

```

### CODE REVIEW

File: datafusion/sql/src/unparser/plan.rs [hunk 1]
```diff
@@ -476,22 +474,72 @@ impl Unparser<'_> {
...
-                    join.left.as_ref(),
+                    right_plan.as_ref(),
                     query,
                     select,
-                    relation,
+                    &mut right_relation,
                 )?;
+
+                let join_filters = if table_scan_filters.is_empty() {
+                    join.filter.clone() // Return a clone of `join.filter` (Option<Expr>) if there are no table scan filters
```
Comment (sgrebnov) [inline datafusion/sql/src/unparser/plan.rs:513]:
```suggestion
                    join.filter.clone()
```

File: datafusion/sql/src/unparser/ast.rs [hunk 2]
```diff
@@ -182,7 +182,20 @@ impl SelectBuilder {
         self
     }
     pub fn selection(&mut self, value: Option<ast::Expr>) -> &mut Self {
-        self.selection = value;
+        match (&self.selection, value) {
```
Comment (sgrebnov) [inline datafusion/sql/src/unparser/ast.rs:185]: This is unrelated but required to support table scan pushdowns. After we enabled push down filters to table scans we have plans where we have dedicated filter as previously and filters as part of table scand that processed separately and override main logical plan filter. This combines all filters to single `where` statement

Comment (sgrebnov) [reply datafusion/sql/src/unparser/ast.rs:185]:
```sql
select
        c_phone as cntrycode,
        c_acctbal
from
        customer
where
        c_mktsegment = 'BUILDING'
        and c_acctbal > (
        select
                avg(c_acctbal)
        from
                customer);
```

```console
|               |   Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
|               |  Projection: customer.c_phone AS cntrycode, customer.c_acctbal                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
|               |   Filter: CAST(customer.c_acctbal AS Decimal128(38, 6)) > (<subquery>)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
|               |     Subquery:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
|               |       Projection: avg(customer.c_acctbal)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
|               |         Aggregate: groupBy=[[]], aggr=[[avg(customer.c_acctbal)]]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|               |           TableScan: customer                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
|               |     TableScan: customer, full_filters=[customer.c_mktsegment = Utf8("BUILDING")] 
```

Comment (sgrebnov) [reply datafusion/sql/src/unparser/ast.rs:185]: I'll add unit test for this case and submit as a separate PR to datafusion main. Combining here to simplify testing/running benchmarks

File: datafusion/sql/src/unparser/plan.rs [hunk 3]
```diff
@@ -476,22 +474,72 @@ impl Unparser<'_> {
...
+                } else {
+                    // Combine `table_scan_filters` into a single filter using `AND`
+                    let combined_filters = table_scan_filters
+                        .into_iter()
+                        .reduce(|acc, filter| Expr::BinaryExpr(BinaryExpr {
+                            left: Box::new(acc),
+                            op: Operator::And,
+                            right: Box::new(filter),
+                        }))
+                        .unwrap();
```
Comment (phillipleblanc) [inline datafusion/sql/src/unparser/plan.rs:523]: Can this at least be an `.expect` with a message on why it shouldn't panic?

File: datafusion/sql/src/unparser/utils.rs [hunk 4]
```diff
@@ -85,6 +85,69 @@ pub(crate) fn find_unnest_node_within_select(
     }
 }
 
+pub (crate) fn try_transform_to_simple_table_scan_with_filters(plan: &LogicalPlan) -> Option<(LogicalPlan, Vec<Expr>)> {
```
Comment (phillipleblanc) [inline datafusion/sql/src/unparser/utils.rs:88]: Add comments explaining what this function does

**Review by phillipleblanc — APPROVED** [review]

File: datafusion/sql/src/unparser/utils.rs (hunk 4 above)
Comment (sgrebnov) [reply datafusion/sql/src/unparser/utils.rs:104]: Added - thank you

```
/// Iterates through the children of a [LogicalPlan] to find a TableScan node before encountering
/// a Projection or any unexpected node that indicates the presence of a Projection (SELECT) in the plan.
/// If a TableScan node is found, returns the TableScan node without filters, along with the collected filters separately.
/// If the plan contains a Projection, returns None.
/// 
/// Note: If a table alias is present, TableScan filters are rewritten to reference the alias.
///
/// LogicalPlan example:
///   Filter: ta.j1_id < 5 
///     Alias:  ta
///       TableScan: j1, j1_id > 10 
///
/// Will return LogicalPlan below:
///     Alias:  ta
///       TableScan: j1
/// And filters: [ta.j1_id < 5, ta.j1_id > 10]
```

File: datafusion/sql/src/unparser/plan.rs (hunk 3 above)
Comment (sgrebnov) [reply datafusion/sql/src/unparser/plan.rs]: Fixed - thank you

**Review by Sevenannn — APPROVED** [review]

**Pull request merged**